### PR TITLE
[REFACTOR][RUNTIME] Update NDArray use the Unified Object System

### PR DESCRIPTION
**Pull request by tqchen**
RFC: https://github.com/apache/incubator-tvm/issues/4286
Previously NDArray has its own object reference counting mechanism.
This PR migrates NDArray to the unified object protocol.

The calling convention of NDArray remained intact.
That means NDArray still has its own type_code and
its handle is still DLTensor compatible.

In order to do so, this PR added a few minimum runtime type
detection in TVMArgValue and RetValue only when the corresponding
type is a base type(ObjectRef) that could also refer to NDArray.

This means that even if we return a base reference object ObjectRef
which refers to the NDArray. The type_code will still be translated
correctly as kNDArrayContainer.
If we assign a non-base type(say Expr) that we know is not compatible
with NDArray during compile time, no runtime type detection will be performed.

This PR also adopts the object protocol for NDArray sub-classing and
removed the legacy NDArray subclass protocol.
Examples in apps/extension are now updated to reflect that.

Making NDArray as an Object brings all the benefits of the object system.
For example, we can now use the Array container to store NDArrays.

### CODE REVIEW

**Comment by tqchen** [comment]
This PR depends on #4578 

cc @wweic @jroesch @junrushao1994 @yzhliu @icemelon9 

After this PR, we can safely remove vm::Tensor and use NDArray directly. Note that however, the type_code for NDArray is still going to be different from the normal Objects during Packed calls. This fact is only relevant if we want to write new codegen that interacts with the system.

Followup discussions can be made to see if we want to further merge the calling convention, there are some pros and cons.

**Comment by tqchen** [comment]
@FrozenGene can you help review this as well?

**Comment by tqchen** [comment]
This PR is now ready to be reviewed

**Comment by junrushao1994** [comment]
So do we really need such an example? Given the fact that NDArrays have been treated almost the same as other tvm objects

**Comment by tqchen** [comment]
given that the example existed historically, i feel it is better to keep it as it is :)

File: apps/extension/src/tvm_ext.cc [hunk 1]
```diff
@@ -65,41 +52,45 @@ class NDSubClass : public tvm::runtime::NDArray {
...
   class SubContainer : public NDArray::Container {
    public:
-    SubContainer(int addtional_info) :
-      addtional_info_(addtional_info) {
-      array_type_code_ = array_type_info<NDSubClass>::code;
-    }
-    static bool Is(NDArray::Container *container) {
-      SubContainer *c = static_cast<SubContainer*>(container);
-      return c->array_type_code_ == array_type_info<NDSubClass>::code;
+    SubContainer(int additional_info) :
```
Comment (wweic) [inline apps/extension/src/tvm_ext.cc]: The comment between line 47 and line 49 needs to be updated.

File: include/tvm/runtime/ndarray.h [hunk 2]
```diff
@@ -191,36 +144,40 @@ class NDArray {
...
+  /*!
+   * \brief Get mutable internal container pointer.
+   * \return a mutable container pointer.
+   */
+  inline Container* get_mutable() const;
+  // Helper functions for FFI handling.
+  /*!
+   * \brief Construct NDArray's Data field from array handle in FFI.
+   * \param handle The array handle.
+   * \return The constructed NDArray.
```
Comment (wweic) [inline include/tvm/runtime/ndarray.h]: The return type does not match `ObjectPtr`.

File: include/tvm/runtime/ndarray.h [hunk 3]
```diff
@@ -259,42 +211,27 @@ class NDArray::Container {
...
-
   /*!
    * \brief The shape container,
    *  can be used used for shape data.
    */
   std::vector<int64_t> shape_;
+};
 
+/*!
+ * \brief Object container class taht backs NDArray.
```
Comment (wweic) [inline include/tvm/runtime/ndarray.h]:
```suggestion
 * \brief Object container class that backs NDArray.
```

File: include/tvm/runtime/ndarray.h [hunk 4]
```diff
@@ -315,49 +254,38 @@ class NDArray::Container {
...
-  : data_(other.data_) {
-  if (data_ != nullptr) {
-    data_->IncRef();
-  }
-}
+  // Information for object protocol.
+  static constexpr const uint32_t _type_index = TypeIndex::kDynamic;
+  static constexpr const uint32_t _type_child_slots = 0;
+  static constexpr const uint32_t _type_child_slots_can_overflow = true;
+  static constexpr const char* _type_key = "NDArray";
```
Comment (wweic) [inline include/tvm/runtime/ndarray.h]: Should this be `"Container"`?

File: include/tvm/runtime/ndarray.h [hunk 5]
```diff
@@ -315,49 +254,38 @@ class NDArray::Container {
...
-    data_ = nullptr;
-  }
-}
+ protected:
+  friend class RPCWrappedFunc;
+  friend class NDArray;
+};
+
+// implementations of inline functions
+// the usages of functions are documented in place.a
```
Comment (wweic) [inline include/tvm/runtime/ndarray.h]:
```suggestion
// the usages of functions are documented in place.
```

File: include/tvm/runtime/ndarray.h (hunk 4 above)
Comment (tqchen) [reply include/tvm/runtime/ndarray.h]: We should refer to the final reference type name, so NDArray is correct here

**Comment by tqchen** [comment]
Thanks @wweic I have updated per your comments

**Comment by tqchen** [comment]
Thanks @wweic @ZihengJiang @junrushao1994

**Comment by tqchen** [comment]
@wweic @jroesch now that this is merged. We should start a refactor PR to remove relay.vm.Tensor and directly use NDArray

**Pull request merged**